src/data_structures/single_linked_list.rs

```diff
@@ -0,0 +1,248 @@
+// the public struct can hide the implementation detail
+pub struct SingleLinkedList<T> {
+    head: Link<T>,
+}
+
+type Link<T> = Option<Box<Node<T>>>;
+
+struct Node<T> {
+    elem: T,
+    next: Link<T>,
+}
+
+impl<T> SingleLinkedList<T> {
+    // Self is an alias for SingleLinkedList
+    // We implement associated function name new for single-linked-list
+    pub fn new() -> Self {
+        // for new function we need to return a new instance
+        SingleLinkedList {
+            // we refer to variants of an enum using :: the namespacing operator
+            head: None,
+        } // we need to return the variant, so there without the ;
+    }
+
+    // As we know the primary forms that self can take: self, &mut self and &self, push will change the linked list
+    // so we need &mut
+    // The push method which the signature's first parameter is self
+    pub fn push(&mut self, elem: T) {
+        let new_node = Box::new(Node {
+            elem,
+            // mem::replace function lets us steal a value out of a borrow by replacing it with another value
+            next: self.head.take(),
+        });
+
+        self.head = Some(new_node);
+    }
+    ///
+    /// In pop function, we trying to:
+    /// * check if the list is empty, so we use enum Option<T>, it can either be Some(T) or None
+    ///   * if it's empty, return None
+    ///   * if it's not empty
+    ///     * remove the head of the list
+    ///     * remove its elem
+    ///     * replace the list's head with its next
+    ///     * return Some(elem), as the situation if need
+    ///
+    /// so, we need to remove the head, and return the value of the head
+    pub fn pop(&mut self) -> Option<T> {
```

Comment:
Wouldn't it be better to return `Result<T, UnderflowError>` and error when the list/stack is empty?

Reply:
Agree with you, but here `pop` return `Option<T>` to support we implement `next(&mut self) -> Option<Self::Item>` method `trait Iterator` at below. And we can do this on `peek` function.

So, please help take a review. thanks.
```
    pub fn peek(&self) -> Result<&T, &str> {
        // Converts from &Option<T> to Option<&T>.
        match self.head.as_ref() {
            None => Err("Stack is empty"),
            Some(node) => Ok(&node.elem),
        }
    }
```

Reply:
* `pop()` [should](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)#Linked_list) really error on empty stacks. If you want to use it in iteration, you can deal with error handling there, converting `UnderflowError` to `None`.
* `peek()` may or may not error depending on implementation. But it's better to be consistent and do whatever is chosen with `peek_mut()` as well.

Reply:
No problem, the `pop()` using `Result<T, Err>` and return `Err` while the list empty. And the `peek()` we follow the style like `peek_mut()`